File: medium/1026. Maximum Difference Between Node and Ancestor/main.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    int maxAncestorDiff(TreeNode* root) {
        int maxDiff = 0;

        function<void(TreeNode*, int, int)> dfs = [&](TreeNode* node, int minAncestor, int maxAncestor) {
            if (!node) return;

            maxDiff = max(maxDiff, abs(node->val - minAncestor));
            maxDiff = max(maxDiff, abs(node->val - maxAncestor));

            minAncestor = min(minAncestor, node->val);
            maxAncestor = max(maxAncestor, node->val);

            dfs(node->left, minAncestor, maxAncestor);
            dfs(node->right, minAncestor, maxAncestor);
        };

        dfs(root, root->val, root->val);
        return maxDiff;
    }
};

File: medium/1026. Maximum Difference Between Node and Ancestor/main.cs
public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution {
    public int MaxAncestorDiff(TreeNode root) {
        int maxDiff = 0;

        void Dfs(TreeNode node, int minAncestor, int maxAncestor) {
            if (node == null) return;

            maxDiff = Math.Max(maxDiff, Math.Abs(node.val - minAncestor));
            maxDiff = Math.Max(maxDiff, Math.Abs(node.val - maxAncestor));

            minAncestor = Math.Min(minAncestor, node.val);
            maxAncestor = Math.Max(maxAncestor, node.val);

            Dfs(node.left, minAncestor, maxAncestor);
            Dfs(node.right, minAncestor, maxAncestor);
        }

        Dfs(root, root.val, root.val);
        return maxDiff;
    }
}
